### ISSUE TITLE
fp8e4m3 / fp8e5m2 -> bf16 conversion silently decodes NaN/Inf as finite values

### ISSUE THREAD

**Issue by etiotto** [intel/intel-xpu-backend-for-triton#7771]
### Summary

The FP8 → BF16 converters in `third_party/intel/lib/TritonIntelGPUToLLVM/ElementwiseOpToLLVM.cpp` decode the reserved, non-finite fp8 encodings as finite bf16 values instead of propagating NaN/Inf. A NaN or Inf input is silently turned into a large finite number, with no diagnostic.

- **FP8E4M3FN → BF16:** the two reserved NaN encodings (bytes `0x7F`, `0xFF`) decode to ±480.0.
- **FP8E5M2 → BF16:** the Inf encodings (`0x7C`, `0xFC`) and the NaN encodings (`0x7D`-`0x7F`, `0xFD`-`0xFF`) all decode to finite values.

Both output paths are affected for each format:

- `Fp8E4M3Nv_to_Bf16` / `Fp8E5M2_to_Bf16` — the `fmul` rebias path, used on all non-LTS drivers
- `Fp8E4M3Nv_to_Bf16Table` / `Fp8E5M2_to_Bf16Table` — the LTS lookup-table path

The `→ FP16` converters in the same file do **not** have this problem, so the fate of a non-finite fp8 input currently depends on which output dtype you cast to. Any kernel relying on NaN/Inf propagation through an fp8 → bf16 cast gets a wrong answer. NaN-based masking is the common case: an element intended to be NaN becomes a finite value and participates in the computation. Because the substituted values are finite and moderate in magnitude, the result looks plausible rather than obviously broken.

### Reproducer

```python
import torch, triton, triton.language as tl

@triton.jit
def cast(src, dst, N: tl.constexpr):
    i = tl.arange(0, N)
    tl.store(dst + i, tl.load(src + i).to(dst.dtype.element_ty))

def run(fp8_dtype, bytes_):
    src = torch.tensor(bytes_, dtype=torch.uint8, device="xpu")
    dst = torch.empty(len(bytes_), dtype=torch.bfloat16, device="xpu")
    cast[(1,)](triton.reinterpret(src, fp8_dtype), dst, N=len(bytes_))
    torch.xpu.synchronize()
    bits = dst.view(torch.uint16).tolist()
    for i, byte in enumerate(bytes_):
        v = dst[i]
        print(f"byte 0x{byte:02X} -> 0x{bits[i]:04x}  value={float(v.item()):<12} "
              f"isnan={v.isnan().item()}  isinf={v.isinf().item()}")

print("fp8e4m3 -> bf16:")
run(torch.float8_e4m3fn, [0x7F, 0xFF])
print("fp8e5m2 -> bf16:")
run(torch.float8_e5m2, [0x7C, 0x7D, 0x7E, 0x7F, 0xFC, 0xFF])
```

Actual output on Intel Arc B580 (BMG):

```
fp8e4m3 -> bf16:
byte 0x7F -> 0x43f0  value=480.0        isnan=False  isinf=False   (should be NaN)
byte 0xFF -> 0xc3f0  value=-480.0       isnan=False  isinf=False   (should be NaN)
fp8e5m2 -> bf16:
byte 0x7C -> 0x4780  value=65536.0      isnan=False  isinf=False   (should be +Inf)
byte 0x7D -> 0x47a0  value=81920.0      isnan=False  isinf=False   (should be NaN)
byte 0x7E -> 0x47c0  value=98304.0      isnan=False  isinf=False   (should be NaN)
byte 0x7F -> 0x47e0  value=114688.0     isnan=False  isinf=False   (should be NaN)
byte 0xFC -> 0xc780  value=-65536.0     isnan=False  isinf=False   (should be -Inf)
byte 0xFF -> 0xc7e0  value=-114688.0    isnan=False  isinf=False   (should be NaN)
```

Note that the existing golden tests (`python/test/unit/intel/test_fp8e4m3_to_bf16.py`, `test_fp8e5m2_to_bf16.py`) assert these finite values and their docstrings recorded the behavior as intentional, so this was pinned rather than accidentally regressed.

**Comment by etiotto** [comment]
## Root cause and fix

### Root cause

In OCP FP8 E4M3 (`e4m3fn`) the encoding `S.1111.111` is reserved for NaN; the format has no infinities, and its largest finite value is 448.0. Both converters treat that bit pattern as an ordinary normal number and rebias it arithmetically:

```
1.875 x 2^(15-7) = 1.875 x 256 = 480.0
```

480.0 is not even representable in the source format, which is a useful tell that the encoding is being misinterpreted rather than merely rounded.

Per path:

- **`Fp8E4M3Nv_to_Bf16`** (`fmul` path) — the reserved byte survives the mask/shift into the bf16 exponent+mantissa field and is then scaled by the rebias multiply like any normal value.
- **`Fp8E4M3Nv_to_Bf16Table`** (LTS LUT path) — `d = c + 0x3C000000` yields `0x43F00000`, whose high half is bf16 `0x43F0` = 480.0.

oneDNN's overflow-to-Inf detection trick (the one applicable to the fp16 path, where multiplying by 36864.0 makes only the reserved byte overflow) **cannot** be reused here: bf16 shares fp32's 8-bit exponent range, so a rebias multiply can never overflow to Inf. The LUT path contains no `fmul` at all.

### Fix

Add an integer `icmp`+`select` against the pre-multiply magnitude field, substituting the canonical bf16 quiet NaN `0x7FC0` with the sign OR'd back in (so the NaN sign is preserved).

The detector is exact — the masked magnitude equals the sentinel for the two reserved bytes and for no others:

| Path | Detector | NaN constant |
|---|---|---|
| `Fp8E4M3Nv_to_Bf16` | `icmp eq (magnitude & 0x07F0), 0x07F0` on i16 | `0x7FC0` |
| `Fp8E4M3Nv_to_Bf16Table` | `icmp eq (c & 0x07F00000), 0x07F00000` on i32 | `0x7FC00000` |

Cost: 2 extra ops per lane in the `fmul` path, 3 per element in the table path. The `fmul` path must keep its per-lane i16 domain — the existing comment there warns that shape is required to dodge an IGC flush-to-zero bug.

Validated by extending `python/test/unit/intel/test_fp8e4m3_to_bf16.py` (exhaustive, all 256 byte values, bit-exact) and `test/Conversion/intel/fp8e4m3_to_bf16_fma.mlir`. The bf16 NaN is built purely from integer bit operations, so unlike a hardware-generated NaN its payload is architecture-independent and can be asserted as an exact bit pattern.
